test/distributed/_tensor/test_dtensor_compile.py

```diff
@@ -0,0 +1,81 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates
+# Owner(s): ["oncall: distributed"]
+
+import copy
+
+import torch
+import torch._dynamo
+import torch.nn as nn
+from torch.distributed._tensor import DeviceMesh
+from torch.distributed.fsdp import FullyShardedDataParallel as FSDP
+from torch.distributed.tensor.parallel import PairwiseParallel, parallelize_module
+from torch.distributed.tensor.parallel.fsdp import enable_2d_with_fsdp
+from torch.testing._internal.common_utils import run_tests
+from torch.testing._internal.distributed._tensor.common_dtensor import (
+    DTensorTestBase,
+    MLPModule,
+    with_comms,
+)
+
+
+class SimpleModel(nn.Module):
+    def __init__(self, device):
+        super().__init__()
+        self.mlp_0 = MLPModule(device)
+        self.mlp_1 = MLPModule(device)
+
+    def forward(self, input):
+        return self.mlp_1(self.mlp_0(input))
+
+
+class TestDTensorCompile(DTensorTestBase):
+    @property
+    def world_size(self):
+        return 4
+
+    @with_comms
+    def test_2d_fsdp_tp_compile(self):
+        data_parallel_size = 2
+        model = SimpleModel(self.device_type)
+        model_copy = copy.deepcopy(model)
+        enable_2d_with_fsdp()
+
+        # 2-D mesh is [dp, tp]
+        twod_mesh = DeviceMesh(
+            device_type="cuda",
+            mesh=torch.arange(0, self.world_size).view(data_parallel_size, -1),
+        )
+
+        fsdp_pg = twod_mesh.get_dim_groups()[0]
+
+        inp = torch.rand(20, 10, device=self.device_type)
+        tp_model = parallelize_module(
+            model, twod_mesh, PairwiseParallel(), tp_mesh_dim=1
+        )
+        eager_2d = FSDP(
+            tp_model, process_group=fsdp_pg, device_id=self.rank, use_orig_params=True
+        )
+        out = eager_2d(inp)
+        # TODO: once aot autograd support is ready we can just use default backend
+        tp_model2 = parallelize_module(
+            model_copy, twod_mesh, PairwiseParallel(), tp_mesh_dim=1
+        )
+        compiled_tp = torch.compile(tp_model2, backend="eager", fullgraph=True)
```

Comment:
cc @wconstab this is mentioned ordering issue currently, the graph break based FSDP somehow does not work with dtensor so I have to put the torch.compile before FSDP wrapping, I double checked this would still generate the correct TP graph for us.

This is actually a bit bad UX since we can't put torch.compile at the end, I'll do some debugging and see if the current graph break FSDP could work with dtensor.